### PR TITLE
OAuth 2.0 BETA branch

### PR DESCRIPTION
**Pull request by larsjohansen-okta**
This is the beta-doc PR for OAuth 2.0. We will use it to incorporate all doc changes until we beta private resource server. The list below keeps track of the sections added/modified compared to the regular doc.

- Added OAuth 2.0 beta page
- Added custom scopes and claims to the beta page
- Added service client and client credentials flow

Reviewers: @federations-okta 

Note: We don't intend to merge this until we actually beta OAuth Private Resource Server.

### CODE REVIEW

File: _source/docs/api/resources/oauth2-beta.md [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+
+> This feature is currently in **Beta** status.
+
+The admin can configure custom scopes and claims via the UI. When custom scopes and claims are configured for a client, the access tokens that are minted will include the custom claims. It will also include custom scopes if they were part of the request.
+
+#### Custom scopes
+If the request that generates the access token contains any custom scopes, those scopes will be part of the <b>scp</b> claim together with the scopes provided from the [OIDC specification](http://openid.net/specs/openid-connect-core-1_0.html).
+
+#### Custom claims
+All configured custom claims will be part of the access token and evaluated for the user.
```
Comment (cwu-okta) [inline _source/docs/api/resources/oauth2-beta.md]: Should you indicate where/how you can configure custom scopes and claims?

File: _source/docs/api/resources/oauth2-beta.md [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+token_type_hint | A hint of the type of <em>token</em>.                                                               | String     |
+client_id       | The client ID generated as a part of client registration. This is used in conjunction with the <em>client_secret</em> parameter to authenticate the client application. | String |
+client_secret   | The client secret generated as a part of client registration. This is used in conjunction with the <em>client_id</em> parameter to authenticate the client application. | String |
+
+##### Token Authentication Methods
+
+A client may only revoke a token generated for that client.
+
+The client can authenticate by providing <em>client_id</em> and <em>client_secret</em> as a part of the URL-encoded form parameters (as described in table above),
+or it can use basic authentication by providing the <em>client_id</em> and <em>client_secret</em> as an Authroization header using the Basic auth scheme.
```
Comment (cwu-okta) [inline _source/docs/api/resources/oauth2-beta.md]: Authroization -> Authorization

File: _source/docs/api/resources/oauth2-beta.md [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+exp         | The expiration time of the token in seconds since January 1, 1970 UTC.                              | long       |
+iat         | The issuing time of the token in seconds since January 1, 1970 UTC.                                 | long       |
+nbf         | A timestamp in seconds since January 1, 1970 UTC when this token is not be used before.             | long       |
+sub         | The subject of the token.                                                                           | String     |
+aud         | The audience of the token.                                                                          | String     |
+iss         | The issuer of the token.                                                                            | String     |
+jti         | The identifier of the token.                                                                        | String     |
+device_id   | The ID of the device associated with the token                                                      | String     |
+
+Custom scopes are added to the list of scopes. Note that this feature is in Beta status.
```
Comment (cwu-okta) [inline _source/docs/api/resources/oauth2-beta.md]: Should there be an example of the custom scopes returned in a request?

Comment (royalchan-okta) [inline _source/docs/api/resources/oauth2-beta.md]: They are just string values, maybe add an extra Response Example for custom scopes so that you can compare

**Comment by royalchan-okta** [comment]
Not related to your change but since you are fixing some `oauth2.md` bugs, the <em> on line 132 and 140 in the response example should be removed

screen shot 2016-06-08 at 10 22 11 pm

File: _source/docs/api/resources/oauth2-beta.md [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+redirect_uri       | Expected if grant_type is <em>authorization_code</em>. Specifies the callback location where the authorization was sent; must match what is preregistered in Okta for this client. | String |
+client_id          | The client ID generated as a part of client registration. This is used in conjunction with the <em>client_secret</em> parameter to authenticate the client application. | String |
+client_secret      | The client secret generated as a part of client registration. This is used in conjunction with the <em>client_id</em> parameter to authenticate the client application. | String |
+code_verifier      | The code verifier of [PKCE](#parameter-details). Okta uses it to recompute the code_challenge and verify if it matches the original code_challenge in the authorization request. | String |
+
+
+##### Token Authentication Methods
+
+The client can authenticate by providing <em>client_id</em> and <em>client_secret</em> as a part of the URL-encoded form parameters (as described in table above),
+or it can use basic authentication by providing the <em>client_id</em> and <em>client_secret</em> as an Authroization header using the Basic auth scheme.
```
Comment (royalchan-okta) [inline _source/docs/api/resources/oauth2-beta.md]: Authroization

**Comment by royalchan-okta** [comment]
since this is a beta branch, you should just add oauth2-beta.md to the docs.yml so you can preview the html

**Comment by mystiberry-okta** [comment]
I'm confused. We already have http://developer.okta.com/docs/api/resources/oauth2.html published as EA. Stephen Lee was asking about it, too. Are we repeating content in the two docs to make it easy for the beta customers? Just want to make sure everything's okay...

File: _source/docs/api/resources/oauth2-beta.md (hunk 4 above)
Comment (mystiberry-okta) [inline _source/docs/api/resources/oauth2-beta.md]: I'll find it and fix in oauth2.md

**Comment by royalchan-okta** [comment]
@mystiberry-okta This is the actual beta branch that we originally discussed for the purposes of doc'ing all the OAuth changes that aren't EA yet.

**Comment by cwu-okta** [comment]
@mystiberry-okta this is a branch which will make available when oauth private resource server goes to beta (targeted for end of June)

**Comment by mystiberry-okta** [comment]
So we'll still have one OAuth 2.0 file when we're done, yes?

**Comment by cwu-okta** [comment]
When we are in oauth private resource server beta there will be 2 OAuth2.0 files. Once we go into OAuth private resource server EA there will be a single OAuth 2.0 file.

**Comment by mystiberry-okta** [comment]
Perfect. We just have to be careful about catching any changes in the older file for the duplicated content in the newer file. Thanks!

**Comment by larsjohansen-okta** [comment]
@royalchan-okta we cannot add it to the ```docs.yml``` since when we merge this (when private resource server goes beta) we don't want the beta page to show up on the page other than if you are provided the link directly.

File: _source/docs/api/resources/oauth2-beta.md (hunk 1 above)
Comment (larsjohansen-okta) [inline _source/docs/api/resources/oauth2-beta.md]: Added it to the "Custom scopes and claims" section.

File: _source/docs/api/resources/oauth2-beta.md (hunk 2 above)
Comment (larsjohansen-okta) [inline _source/docs/api/resources/oauth2-beta.md]: Fixed.

**Comment by larsjohansen-okta** [comment]
@royalchan-okta could you explain what should be removed in the example you took a screenshot of?

File: _source/docs/api/resources/oauth2-beta.md (hunk 4 above)
Comment (larsjohansen-okta) [inline _source/docs/api/resources/oauth2-beta.md]: Fixed it here as well.

**Comment by larsjohansen-okta** [comment]
image

**Comment by royalchan-okta** [comment]
@larsjohansen-okta the \<em\> should be removed, it actually turned it into markup in github and italicized the words...

**Comment by larsjohansen-okta** [comment]
@royalchan-okta Gotcha. I fixed that in the beta-doc, but forgot to fix it in oauth2.md. It's now fixed in both. It was also missing a ```&``` between token_type and state parameter.

File: _source/docs/api/resources/oauth2-beta.md [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+
+Error Id                |  Details                                                                                                     |
+------------------------+--------------------------------------------------------------------------------------------------------------|
+invalid_client          | The specified client id wasn't found. |
+invalid_request         | The request structure was invalid. E.g. the basic authentication header was malformed, or both header and form parameters were used for authentication or no authentication information was provided. |
+
+#### Response Example (Success)
+
+~~~http
+HTTP/1.1 200 OK
```
Comment (royalchan-okta) [inline _source/docs/api/resources/oauth2-beta.md]: I forgot to fix this, but it turns out this returns a `204 No Content`

**Comment by royalchan-okta** [comment]
LGTM, just that one minor fix left

File: _source/docs/api/resources/oauth2-beta.md (hunk 5 above)
Comment (larsjohansen-okta) [inline _source/docs/api/resources/oauth2-beta.md]: Fixed.

**Comment by mystiberry-okta** [comment]
Since we can't push this until ready for Beta, the oauth2.md fixes won't get pushed either. Unless there's some git ninja tricks for only pushing one file?

**Comment by larsjohansen-okta** [comment]
@mystiberry-okta you're right, and no there's no way we can push this single file. We can create a new branch and copy the changes for this single file. I suggest we do that so we get the bug fixes in as soon as possible, so I created a PR with the changes for oauth2.md only here https://github.com/okta/okta.github.io/pull/270. Please review so we can merge those typos and minor bug fixes.

File: _source/docs/api/resources/oauth-clients.md [hunk 6]
```diff
@@ -63,10 +63,10 @@ Client applications have the following properties:
 | client_name                | human-readable string name of the client application              | String                                                                 | FALSE    | FALSE  | FALSE    |
 | client_secret              | OAuth 2.0 client secret string (used for confidential clients)    | String                                                                 | TRUE     | TRUE   | TRUE     |
 | logo_uri                   | URL string that references a logo for the client                  | String                                                                 | TRUE     | FALSE  | FALSE    | 
-| application_type           | The type of client application                                    | `web`, `native`, or `browser`                                          | TRUE     | TRUE   | TRUE     |
+| application_type           | The type of client application                                    | `web`, `native`, `browser` or `service`                                | TRUE     | TRUE   | TRUE     |
```
Comment (larsjohansen-okta) [inline _source/docs/api/resources/oauth-clients.md]: Do we need to doc that the service app type is in beta once we merge this PR?

File: _source/docs/api/resources/oauth2-beta.md [hunk 7]
```diff
@@ -262,7 +265,7 @@ The admin can configure custom scopes and claims via the UI <a href="...link to
 If the request that generates the access token contains any custom scopes, those scopes will be part of the <b>scp</b> claim together with the scopes provided from the [OIDC specification](http://openid.net/specs/openid-connect-core-1_0.html).
 
 #### Custom claims
-All configured custom claims will be part of the access token and evaluated for the user.
+All configured custom claims will be part of the access token. And they will be evaluated for the user if the grant type is not <em>client_credentials</em>.
```
Comment (larsjohansen-okta) [inline _source/docs/api/resources/oauth2-beta.md]: I would rephrase it like this: ```All configured custom claims will be part of the access token. They will be evaluated for the user if the grant type is not <em>client_credentials</em>.```

File: _source/docs/api/resources/oauth-clients.md (hunk 6 above)
Comment (yuliu-okta) [inline _source/docs/api/resources/oauth-clients.md]: Added both here and in the token endpoint.

File: _source/docs/api/resources/oauth2-beta.md (hunk 7 above)
Comment (yuliu-okta) [inline _source/docs/api/resources/oauth2-beta.md]: fixed

File: _source/docs/api/resources/oauth-clients.md (hunk 6 above)
Comment (mystiberry-okta) [inline _source/docs/api/resources/oauth-clients.md]: Doc looks fantastic!

**Comment by royalchan-okta** [comment]
Superseded by https://github.com/okta/okta.github.io/pull/299
